src/api/v1/lobby.rs

```diff
@@ -81,19 +84,24 @@ pub(crate) async fn try_contribute(
 
         info.is_first_ping_attempt = false;
         info.last_ping_time = now;
+
+        uid = info.token.unique_identifier().to_owned();
     }
 
     // Check if there is an existing contribution in progress
     if app_state.participant.is_some() {
         return TryContributeResponse::AnotherContributionInProgress;
     }
 
+    // If this insertion fails, worst case we allow multiple contributions from the same participant
```

Comment:
This comment is a bit ominous and seems to infer that we are vulnerable to a sybil attack.

My understanding is that we only allow people in the lobby to contribute and before they reach the lobby, they need to auth and since we check if they've already contributed when they auth, then it would not be pssible for someone to participate twice. We also don't allow the same two UIDs in the lobby.

is this consistent with what you were thinking?

Reply:
> seems to infer that we are vulnerable to a sybil attack

Not really. Maybe I should reword the comment. Explanation below, let me know what you think.

> My understanding is that we only allow people in the lobby to contribute

Correct.

> before they reach the lobby, they need to auth and [...] we check if they've already contributed when they auth

Correct.

> then it would not be pssible for someone to participate twice

Not necessarily correct, this is what this comment is about. :)

We only commit participation info to the db when participants become active contributors (i.e. this line of code). At this point, they passed all the checks and are good to contribute *this time*.

However, this db insertion can fail (db full, connection pool exhausted or whatever) and we will simply "forget" to blacklist them _this time_. When they successfully contribute, they will be logged out as expected. However, since our db had a little fluke, we haven't marked their participation and they can login again.

This opens us up for one person contributing multiple times only when the database is down (i.e. server error) at this particular moment (and the db needs to be fixed when they try to sign in).

Basically, it's such an extreme case that I thought it might be easier to allow someone twice than introduce another error code.

Reply:
Right I see, thats not too bad I think. If our DB is broken in such a simple way, we would probably need to do some maintenance on it and stop the server anyways (unless it was a fluke)